### ISSUE TITLE
Firebase deploy stuck at "cleaning up build files..."

### ISSUE THREAD

**Issue by Wian-TMC** [firebase/firebase-tools#3694]
### [REQUIRED] Environment info

**firebase-tools:** 9.16.0

**Platform:**macOS

### [REQUIRED] Test case

Deploy functions using firebase deploy --only functions

### [REQUIRED] Steps to reproduce

Deploy functions using firebase deploy --only functions

### [REQUIRED] Expected behavior

The deploy should complete after the logs ``` functions: cleaning up build files...```

### [REQUIRED] Actual behavior

The logs are stuck at ``` functions: cleaning up build files...``` and requires a ```Ctrl+C``` to regain terminal.

The above behaviour does not occur everytime. Is it important for it to finish, or can I ignore the fact that it gets stuck and continue without having to try and re-deploy?

**Comment by inlined** [comment]
It shouldn't be getting stuck, though it's not fatal if you ctrl + c to kill that part of the process; it happens after your functions are already done deploying. Out of curiosity, let's try absolutely cleaning your artifacts and seeing if you continue to get stuck afterwards. Try running the following commands:

```
# Get the latest version of the firebase command
npm install -g firebase-tools

# Unlock the deletegcfartifacts command
firebase --open-sesame deletegcfartifacts

# Run a full purge of your GCF artifacts directories
firebase functions:deletegcfartifacts
```

This process will check your per-metro Google Container Repository buckets for the `gcf` folder and recursively delete its contents. *If you've been putting your own images in that directory they will be deleted by this command as well*. Can you run this command to completion and then write back in a week or so whether you continue to see deploys get stuck?

**Comment by Wian-TMC** [comment]
@inlined Will do! Thank you

**Comment by Wian-TMC** [comment]
Still gets stuck. But only when we deploy all our functions at once. Deploying them individually (typically when only one function has changed), it works fine.

Releasing them all at once generates a warning of ```functions: got "Quota Exceeded" error while trying to update [FUNCTION_NAME]. Waiting to retry...``` It retries and successfully deploys these functions as well though, but then it get stuck at ```functions: cleaning up build files...```

**Comment by inlined** [comment]
Can you run your deploy with the --debug flag and send the last few dozen log lines? Ideally that'll show us what sort of requests the SDK is locking up on.

**Comment by Wian-TMC** [comment]
Sure. Here are the last few lines...

```
[2021-08-25T14:57:30.824Z] <<< [apiv2][status] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/f3375f48-d631-460f-847e-656a4925fa0a/manifests/removeOlderNotificationsCron_version-22 202
[2021-08-25T14:57:30.824Z] <<< [apiv2][body] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/f3375f48-d631-460f-847e-656a4925fa0a/manifests/removeOlderNotificationsCron_version-22 {"errors":[]}
[2021-08-25T14:57:30.824Z] >>> [apiv2][query] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/f3375f48-d631-460f-847e-656a4925fa0a/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 [none]
[2021-08-25T14:57:31.237Z] Failed docker command with error Failed to make request to https://us.gcr.io/v2/get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c/tags/list {"name":"FirebaseError","children":[],"exit":1,"message":"Failed to make request to https://us.gcr.io/v2/get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c/tags/list","original":{"message":"request to https://us.gcr.io/v2/get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c/tags/list failed, reason: socket hang up","type":"system","errno":"ECONNRESET","code":"ECONNRESET"},"status":500}
[2021-08-25T14:57:31.347Z] >>> [apiv2][query] GET https://us.gcr.io/v2/get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c/tags/list [none]
[2021-08-25T14:57:32.100Z] <<< [apiv2][status] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/ffaa3f73-a26a-45df-90ff-8427a164b14a/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 202
[2021-08-25T14:57:32.100Z] <<< [apiv2][body] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/ffaa3f73-a26a-45df-90ff-8427a164b14a/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 {"errors":[]}
[2021-08-25T14:57:32.233Z] <<< [apiv2][status] GET https://us.gcr.io/v2/get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c/tags/list 200
[2021-08-25T14:57:32.233Z] <<< [apiv2][body] GET https://us.gcr.io/v2/get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c/tags/list {"child":[],"manifest":{},"name":"get-tod/gcf/us-central1/91bc526f-7520-4d0e-a2fc-1c81a3b01e1c","tags":[]}
[2021-08-25T14:57:32.233Z] >>> [apiv2][query] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/d811d589-b9a9-4416-8329-a29fa59b3ad5/manifests/userCreationVodapayListener_version-13 [none]
[2021-08-25T14:57:32.379Z] <<< [apiv2][status] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/f3375f48-d631-460f-847e-656a4925fa0a/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 202
[2021-08-25T14:57:32.379Z] <<< [apiv2][body] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/f3375f48-d631-460f-847e-656a4925fa0a/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 {"errors":[]}
[2021-08-25T14:57:32.415Z] <<< [apiv2][status] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/dc491ebc-a55f-48f4-bbeb-41b54f441652/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 202
[2021-08-25T14:57:32.415Z] <<< [apiv2][body] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/dc491ebc-a55f-48f4-bbeb-41b54f441652/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 {"errors":[]}
[2021-08-25T14:57:33.132Z] <<< [apiv2][status] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/d811d589-b9a9-4416-8329-a29fa59b3ad5/manifests/userCreationVodapayListener_version-13 202
[2021-08-25T14:57:33.132Z] <<< [apiv2][body] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/d811d589-b9a9-4416-8329-a29fa59b3ad5/manifests/userCreationVodapayListener_version-13 {"errors":[]}
[2021-08-25T14:57:33.132Z] >>> [apiv2][query] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/d811d589-b9a9-4416-8329-a29fa59b3ad5/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 [none]
[2021-08-25T14:57:34.727Z] <<< [apiv2][status] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/d811d589-b9a9-4416-8329-a29fa59b3ad5/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 202
[2021-08-25T14:57:34.727Z] <<< [apiv2][body] DELETE https://us.gcr.io/v2/get-tod/gcf/us-central1/d811d589-b9a9-4416-8329-a29fa59b3ad5/manifests/sha256:2c1cdf079239a9e211d0fecc34a7920b0704bca272f6e5372397a9017185aa16 {"errors":[]}
```
Got stuck after the last line above...

**Comment by edcmex** [comment]
got the same issue.

**Comment by edcmex** [comment]
firebase logout and then login fixed it for me.

**Comment by mukarram-rt** [comment]
It is occurring in my build pipeline on Google Cloud Build. It stucks at `functions: cleaning up build files...` and finally timeout error is thrown.

**Comment by mukarram-rt** [comment]
It is constantly occurring on the Cloud Build pipeline. There are some steps after functions deployment and because of this issue, those steps are not getting executed. And cloud build throws timeout error.

**Comment by dansjack** [comment]
> firebase logout and then login fixed it for me.

This worked for me too, thanks!

**Comment by rhodgkins** [comment]
I get stuck on this last bit too. Eventually network stops working on my machine* and then a few minutes later the command exits with a failure. Or I can CMD+C and everything eventually starts working...

* I've only realised it stops working as I was streaming something and it stopped and then I could load any web pages till the command had exited.

**Comment by MaxGenash** [comment]
Got a similar problem in Bitbucket Pipelines after updating `firebase-tools` from 9.10.0 to 9.22.0.
Previously our deployment could normally deploy our 230 cloud functions (even though it said `got "Quota Exceeded" error...` during the deployment process, but could handle it eventually), but now it fails after/during cleaning-up build files:
```
i  functions: cleaning up build files...
✖ Deployment failed.
```
In the debug mode I noticed that there are many messages `[2021-11-17T14:59:54.929Z] Failed docker command with error  Timeout {}`

**Comment by inlined** [comment]
The process of deleting from Google Container Repository is chatty and at 230 functions it failed despite retries. There's two pieces of good news though:

1. Everything about your functions deploy happened before that failure. You don't need to redeploy at all.
2. We're moving off of Google Container Repository and onto Artifact registry. We'll reduce the number of API calls to make by several fold and your deploys should be faster and more reliable. This may take until the end of the year though.

**Comment by MaxGenash** [comment]
@inlined thanks for the response.

Do you have any idea why it works in firebase-tools@9.10.0, but doesn't work in firebase-tools@9.22.0 (the number of functions is the same)?

Actually, about half of the functions aren't deployed, and there is a message "Functions deploy had errors with the following functions:.../*names of about 100 functions*/...". Thought maybe our problem is not related to cleaning up build files, but has some other reasons, e.g. maybe the code of firebase-tools got slower between v 9.10.0 and v9.22.0, so it can no longer handle such amount of functions being deployed in parallel.

**Comment by inlined** [comment]
Goodness. This issue sat in my inbox for an inexcusable amount of time. Sorry; I don't get to code much these days and didn't check GitHub for a few weeks. The code to deploy your functions has been completely rewritten between those two versions. Some changes intended to improve reliability and some also intended to help performance.

Can you file a support ticket at https://firebase.google.com/support to make sure you get the attention you deserve? A firebase-debug.log would let us see pretty quickly where the failures are and help us fix whatever regression is hitting you.

**Comment by axwaxw** [comment]
I am experiencing the same issue.

firebase timeout on deploy with the following errors:

```
i  functions: cleaning up build files...
!  functions: Unhandled error cleaning up build images. This could result in a small monthly bill if not corrected. You can attempt to delete these images by redeploying or you can delete them manually at https://console.cloud.google.com/gcr/images/***t/us/gcf
i  hosting[***]: finalizing version...

Error: Authentication Error: Your credentials are no longer valid. Please run firebase login --reauth
```

The debug log shows:

```  Failed docker command with error Timeout {}  ```

I have manually deleted the relevant images, but this does not make a difference.

I get the same errors every time I attempt to deploy.

This app has been up and deploying successfully for over 18 months.

Any word on a workaround or solution to this issue?  Currently, I am unable to deploy updates....

**Comment by inlined** [comment]
We should probably change the error message you saw to clarify that your functions _have_ deployed correctly; we just weren't able to clean up the leftover images in Google Container Repository. Hardening this deletion code is a bit of a low priority because we're actually in the process of migrating everyone from Container Repository to Artifact Registry (which has better regionalization and a free tier). Expect an update in ~1-2mo on that front.

It's possible that the cleanup is taking so long (and failing) because there's a lot of cruft from old functions before we used to clean up GCR images. We actually have a hidden tool you can use to blow away this cache:

```
# enable the hidden command
firebase --open-sesame deletegcfartifacts
# purge everything
firebase functions:deletegcfartifacts
```

It's possible that this command will fail halfway through as well, but as it makes progress your backlog gets smaller and you will eventually finish. After that, the process of cleaning up your build images should be faster/more reliable

In the meanitme, @axwaxw it _does_ look like your hosting site failed to finalize. If you follow the instructions and run `firebase login --reauth` then your problems should go away.

**Comment by axwaxw** [comment]
Hi Thomas,

Thanks for getting back to me.

Yes – I was aware of the firebase function firebase functions:deletegcfartifacts and I have been using that to purge images before each deployment – however I still get the same error when I deploy.

I think I can deploy the functions and hosting individually without error, but obviously this is less than ideal since I have to babysit the process quite carefully to ensure no errors.

Thanks

Alex

**Comment by inlined** [comment]
I believe there was a major bug, which I hope to fix in #4477 that will fix polynomial growth in API calls in proportion to the number of functions being deployed. The next build should hopefully reduce your build times dramatically.

**Comment by yuchenshi** [comment]
Should be fixed in the latest release.
